src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template

```diff
@@ -75,15 +77,31 @@ class Direct$Type$Buffer$RW$$BO$
 
 #if[byte]
 
-    private record Deallocator(long address, long size, int capacity) implements Runnable {
-        private Deallocator {
+    private static final class Deallocator implements Runnable {
+
+        private static final long INVOKED_OFFSET = Unsafe.getUnsafe()
+                .objectFieldOffset(Deallocator.class, "invoked");
+
+        private final long address;
+        private final long size;
+        private final int capacity;
+        private int invoked;
+
+        private Deallocator(long address, long size, int capacity) {
             assert address != 0;
+            this.address = address;
+            this.size = size;
+            this.capacity = capacity;
         }
 
         public void run() {
-            UNSAFE.freeMemory(address);
-            Bits.unreserveMemory(size, capacity);
+            // Ensure idempotency (paranoia)
+            if (Unsafe.getUnsafe().compareAndSetInt(this, INVOKED_OFFSET, 0, 1)) {
```

Comment:
a question for my edification: is this compound statement suitably atomic to ensure synchronzied serial access to freeMemory block ?  There's no danger of concurrent interleaved execution resulting in inconsistent view of the member variable via the reflective  access ?

Reply:
The CAS operation establishes a happens-before relation between the CAS operation itself and any following operations (such as `freeMemory()`). Also, only one thread will succeed in changing the value from zero to one.

Reply:
The behavior in this PR is now like the intended behavior in Java 21 (and before) but thread-safe. So, by induction, this should be as safe or safer as we had in <= 21. Adding blocking might introduce other problems in a multi-threaded environment.

Reply:
thanks for the replies, so all invocations lead to Atomic::cmpxchg, where some magic happens....I didn't realise that AtomicInteger is basically a wrapper around what has been added to Deallocator